graphs/greedy_best_first.py

```diff
@@ -0,0 +1,162 @@
+"""
+https://en.wikipedia.org/wiki/Best-first_search#Greedy_BFS
+"""
+
+from typing import List, Tuple
+
+grid = [
+    [0, 0, 0, 0, 0, 0, 0],
+    [0, 1, 0, 0, 0, 0, 0],  # 0 are free path whereas 1's are obstacles
+    [0, 0, 0, 0, 0, 0, 0],
+    [0, 0, 1, 0, 0, 0, 0],
+    [1, 0, 1, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 1, 0, 0],
+]
+
+delta = [[-1, 0], [0, -1], [1, 0], [0, 1]]  # up, left, down, right
+
+
+class Node:
+    """
+    >>> k = Node(0, 0, 4, 5, 0, None)
+    >>> k.calculate_heuristic()
+    9
+    >>> n = Node(1, 4, 3, 4, 2, None)
+    >>> n.calculate_heuristic()
+    2
+    >>> l = [k, n]
+    >>> n == l[0]
+    False
+    >>> l.sort()
+    >>> n == l[0]
+    True
+    """
+
+    def __init__(self, pos_x, pos_y, goal_x, goal_y, g_cost, parent):
+        self.pos_x = pos_x
+        self.pos_y = pos_y
+        self.pos = (pos_y, pos_x)
+        self.goal_x = goal_x
+        self.goal_y = goal_y
+        self.g_cost = g_cost
+        self.parent = parent
+        self.f_cost = self.calculate_heuristic()
+
+    def calculate_heuristic(self) -> float:
+        """
+        The heuristic here is the Manhattan Distance
+        Could elaborate to offer more than one choice
+        """
+        dy = abs(self.pos_x - self.goal_x)
+        dx = abs(self.pos_y - self.goal_y)
+        return dx + dy
+
+    def __lt__(self, other):
+        return self.f_cost < other.f_cost
+
+
+class GreedyBestFirst:
+    def __init__(self, start, goal):
+        self.start = Node(start[1], start[0], goal[1], goal[0], 0, None)
+        self.target = Node(goal[1], goal[0], goal[1], goal[0], 99999, None)
+
+        self.open_nodes = [self.start]
+        self.closed_nodes = []
+
+        self.reached = False
+
+        self.path = [(self.start.pos_y, self.start.pos_x)]
+        self.costs = [0]
+
+    def search(self):
+        while self.open_nodes:
```

Comment:
Please add doctests please.  A search method should return a result, not print() a result.